### PR TITLE
NOISSUE - Add SMPP notifier

### PR DESCRIPTION
**Pull request by blokovi**
Signed-off-by: Ivan Milosevic <iva@blokovi.com>

### What does this do?

Adds an implementation of SMPP Notifier (capability of sending SMS notifications).

### CODE REVIEW

File: consumers/notifiers/service.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ var (
 	ErrSave = errors.New("failed to subscription")
 
 	// ErrNotFound indicates a non-existent entity request.
-	ErrNotFound = errors.New("non-existent entity")
+	ErrNotFound = errors.New("non-existent entityty")
```
Comment (manuio) [inline consumers/notifiers/service.go]: Fix `entityty`

File: consumers/notifiers/service_test.go [hunk 2]
```diff
@@ -29,7 +29,8 @@ func newService() notifiers.Service {
 	auth := mocks.NewAuth(map[string]string{exampleUser1: exampleUser1, exampleUser2: exampleUser2, invalidUser: invalidUser})
 	notifier := mocks.NewNotifier()
 	idp := uuid.NewMock()
-	return notifiers.New(auth, repo, idp, notifier)
+	from := ""
```
Comment (manuio) [inline consumers/notifiers/service_test.go]: Mayeb better to don't use an empty string here?

File: consumers/notifiers/api/endpoint_test.go [hunk 3]
```diff
@@ -68,7 +68,8 @@ func newService(tokens map[string]string) notifiers.Service {
 	repo := mocks.NewRepo(make(map[string]notifiers.Subscription))
 	idp := uuid.NewMock()
 	notif := mocks.NewNotifier()
-	return notifiers.New(auth, repo, idp, notif)
+	from := ""
```
Comment (manuio) [inline consumers/notifiers/api/endpoint_test.go]: Mayeb better to don't use an empty string here?

**Review by manuio — CHANGES_REQUESTED** [review]

File: consumers/notifiers/service_test.go (hunk 2 above)
Comment (dborovcanin) [reply consumers/notifiers/service_test.go]: I agree, simply use a string literal.

File: consumers/notifiers/smpp/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+| MF_SMTP_NOTIFY_WRITER_CONFIG_PATH | Database connection SSL mode (disable, require, verify-ca, verify-full) | disable               |
+| MF_SMTP_NOTIFY_DB_SSL_MODE        | Path to the PEM encoded certificate file                                |                       |
+| MF_SMTP_NOTIFY_DB_SSL_CERT        | Path to the PEM encoded key file                                        |                       |
+| MF_SMTP_NOTIFY_DB_SSL_KEY         | Path to the PEM encoded root certificate file                           |                       |
+| MF_SMTP_NOTIFY_DB_SSL_ROOT_CERT   | Users service HTTP port                                                 | 8180                  |
+| MF_SMTP_NOTIFY_HTTP_PORT          | Path to server certificate in pem format                                |                       |
+| MF_SMTP_NOTIFY_SERVER_CERT        | Path to server cert in pem format                                       |                       |
+| MF_SMTP_NOTIFY_SERVER_KEY         | Path to server key in pem format                                        |                       |
```
Comment (dborovcanin) [inline consumers/notifiers/smpp/README.md]: SMTP docs.

File: consumers/notifiers/smpp/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# SMPP Notifier
+
+SMTP Notifier implements notifier for send SMS notifications.
+
+## Configuration
+
+The Subscription service using SMTP Notifier is configured using the environment variables presented in the
+following table. Note that any unset variables will be replaced with their
+default values.
+
```
Comment (dborovcanin) [inline consumers/notifiers/smpp/README.md]: Please fix README.

File: consumers/notifiers/smpp/notifier.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+		destAddrTON:   cfg.DestAddrTON,
+		sourceAddrNPI: cfg.SourceAddrNPI,
+		destAddrNPI:   cfg.DestAddrNPI,
+	}
+	return ret
+}
+
+func (n *notifier) Notify(from string, to []string, msg messaging.Message) error {
+	fmt.Println(to)
+	fmt.Println(len(to))
```
Comment (dborovcanin) [inline consumers/notifiers/smpp/notifier.go]: Remove.

File: docker/addons/smpp-notifier/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      MF_SMPP_NOTIFIER_DB_HOST: smpp-notifier-db
+      MF_SMPP_NOTIFIER_DB_PORT: ${MF_SMPP_NOTIFIER_DB_PORT}
+      MF_SMPP_NOTIFIER_DB_USER: ${MF_SMPP_NOTIFIER_DB_USER}
+      MF_SMPP_NOTIFIER_DB_PASS: ${MF_SMPP_NOTIFIER_DB_PASS}
+      MF_SMPP_NOTIFIER_DB: ${MF_SMPP_NOTIFIER_DB}
+      MF_SMPP_NOTIFIER_PORT: ${MF_SMPP_NOTIFIER_PORT}
+      MF_NATS_URL: ${MF_NATS_URL}
+      MF_JAEGER_URL: ${MF_JAEGER_URL}
+      MF_AUTH_GRPC_URL: ${MF_AUTH_GRPC_URL}
+      MF_AUTH_GRPC_TIMEOUT: ${MF_AUTH_GRPC_TIMEOUT}
```
Comment (dborovcanin) [inline docker/addons/smpp-notifier/docker-compose.yml]: You can group SMPP-related env vars by putting these on top of the `environment` section.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — APPROVED** [review]

File: consumers/notifiers/smpp/notifier.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+		SourceAddrNPI: n.sourceAddrNPI,
+		DestAddrNPI:   n.destAddrNPI,
+		Text:          pdutext.Raw(msg.Payload),
+		Register:      pdufield.NoDeliveryReceipt,
+	}
+	sm, err := n.t.Submit(send)
+	if err != nil {
+		return err
+	}
+	fmt.Println("Message ID:", sm.RespID())
```
Comment (dborovcanin) [inline consumers/notifiers/smpp/notifier.go]: Remove logs.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — DISMISSED** [review]

File: consumers/notifiers/smpp/notifier.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/mainflux/mainflux/consumers/notifiers"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/mainflux/mainflux/pkg/transformers"
+	"github.com/mainflux/mainflux/pkg/transformers/json"
+)
+
+var _ notifiers.Notifier = (*notifier)(nil)
+
+type notifier struct {
+	t             *smpp.Transmitter
```
Comment (drasko) [inline consumers/notifiers/smpp/notifier.go]: I would consider calling this struct field `transmitter`, and one below `transformer`. When it comes to struct fields, I find it easier to read the code - example on one of the lines below `n.t.Submit()` does not tell me much as `n.transmitter.Submit()`.

@nmarcetic @dusanb94 opinions?

**Review by drasko — CHANGES_REQUESTED** [review]

File: consumers/notifiers/smpp/notifier.go (hunk 9 above)
Comment (dborovcanin) [reply consumers/notifiers/smpp/notifier.go]: For me, both are fine. **Transmitter** is a bit more self-explanatory, **t** is shorter. When reading the code, the one will probably need to check what it is in any case.

Comment (nmarcetic) [reply consumers/notifiers/smpp/notifier.go]: Agree @drasko @dusanb94 , readability wins. I prefer to keep variable names short in scopes (like 1-3 letters), not struct fields where they should be obvious.

File: consumers/notifiers/smpp/notifier.go [hunk 10]
```diff
@@ -18,8 +18,8 @@ import (
 var _ notifiers.Notifier = (*notifier)(nil)
 
 type notifier struct {
-	t             *smpp.Transmitter
-	tr            transformers.Transformer
+	transmiter    *smpp.Transmitter
```
Comment (drasko) [inline consumers/notifiers/smpp/notifier.go]: Typo: `transmitter`

**Review by drasko — CHANGES_REQUESTED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**